File: apps/android/app/src/main/java/com/exponential/app/ui/components/SegmentedTrack.kt
package com.exponential.app.ui.components

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.exponential.app.domain.ContextBarSlice
import com.exponential.app.ui.theme.DesignTokens
import com.exponential.app.ui.theme.GlassTokens

// EXP-1051: the STACKED sibling of [UsageTrack] — one track, many slices, and
// the tick marks that say where the compaction floor and the two usage
// thresholds sit. Same height and same track colour as the single-fill bar, so
// the context-window block and the rate-limit windows above it read as one
// instrument. (Web `ContextBar`, desktop `ui::context_layout`, iOS
// `ContextSegmentedTrack`.)

/**
 * The tone vocabulary the contract's layers name ([ContextBarSlice.tone]) in
 * mobile's palette: the avatar hues (the app's ONE generated colour wheel),
 * `neutral` = the foreground at a whisper, and `track` = the track itself —
 * `free` is never a slice, so a `track` tone draws nothing.
 */
fun contextToneColor(tone: String, foreground: Color): Color = when (tone) {
    "green" -> DesignTokens.Avatar.Green
    "yellow" -> DesignTokens.Avatar.Yellow
    "orange" -> DesignTokens.Avatar.Orange
    "violet" -> DesignTokens.Avatar.Violet
    "pink" -> DesignTokens.Avatar.Pink
    "blue" -> DesignTokens.Avatar.Blue
    "red" -> DesignTokens.Avatar.Red
    "teal" -> DesignTokens.Avatar.Teal
    "track" -> GlassTokens.StrokeStrong
    // `neutral` and anything a newer contract names: the foreground, dimmed.
    else -> foreground.copy(alpha = 0.30f)
}

/**
 * A stacked meter: [slices] left to right as percentages OF THE WHOLE track,
 * CLIPPED at 100% (the device's estimates may overshoot what it measured, and
 * a silently rescaled bar would lie about the layer sizes), with a hairline
 * mark at each of [ticks].
 */
@Composable
fun SegmentedTrack(
    slices: List<Pair<Color, Double>>,
    ticks: List<Int>,
    modifier: Modifier = Modifier,
    height: Dp = UsageTrackHeight,
    tickColor: Color = Color.White.copy(alpha = 0.35f),
) {
    Box(
        modifier = modifier
            .fillMaxWidth()
            .height(height)
            .clip(RoundedCornerShape(height / 2)),
    ) {
        Canvas(modifier = Modifier.fillMaxWidth().height(height)) {
            drawRect(color = GlassTokens.StrokeStrong, size = size)
            var x = 0f
            for ((color, percent) in slices) {
                if (x >= size.width) break
                val width = (percent / 100.0).coerceAtLeast(0.0).toFloat() * size.width
                if (width <= 0f) continue
                val drawn = width.coerceAtMost(size.width - x)
                drawRect(color = color, topLeft = Offset(x, 0f), size = Size(drawn, size.height))
                x += drawn
            }
            // The marks last, so a full bar never buries its own thresholds.
            val tickWidth = 1.dp.toPx()
            for (tick in ticks) {
                val at = (tick / 100f).coerceIn(0f, 1f) * size.width - tickWidth / 2f
                drawRect(
                    color = tickColor,
                    topLeft = Offset(at.coerceIn(0f, size.width - tickWidth), 0f),
                    size = Size(tickWidth, size.height),
                )
            }
        }
    }
}

File: apps/ios/ExpUI/Sources/SegmentedTrack.swift
import ExpCore
import SwiftUI

// EXP-1051: the STACKED meter — `AgentUsageTrack`'s rail, filled by several
// layers instead of one. It draws the context-window bar: the layers the
// device attributed, then the conversation, over the free track.
//
// A sibling of `UsageTrack.swift` rather than a mode of it: the usage rail
// takes ONE percentage and a severity tone, this one takes a list of
// contract-toned slices and never colours itself by how full it is. Both wear
// the same capsule and the same track colour, so the two read as one family.
//
// The ×4 twins: web's stacked `Meter`, desktop `ui::context_layout`, Android
// `SegmentedTrack`. The slices come from
// `ContextLayoutPresentation.contextWindowView`, already ordered and already
// percentages OF THE WHOLE WINDOW — which is why nothing here normalises
// them: an overshooting layout CLIPS (the rule the fold is locked on), it does
// not rescale.

public struct SegmentedTrack: View {
    let slices: [ContextBarSlice]
    /// Percent marks drawn over the fill — the compaction floor and the two
    /// usage thresholds.
    var ticks: [Int]
    var height: CGFloat

    public init(slices: [ContextBarSlice], ticks: [Int] = [], height: CGFloat = 8) {
        self.slices = slices
        self.ticks = ticks
        self.height = height
    }

    public var body: some View {
        GeometryReader { geo in
            ZStack(alignment: .leading) {
                // The track IS the `free` row: it is never a slice.
                Capsule()
                    .fill(GlassTokens.strokeStrong)
                HStack(spacing: 0) {
                    ForEach(slices) { slice in
                        Rectangle()
                            .fill(Self.tone(slice.tone))
                            .frame(width: width(slice.percent, in: geo.size.width))
                    }
                    Spacer(minLength: 0)
                }
                ForEach(ticks, id: \.self) { tick in
                    Rectangle()
                        .fill(Color.black.opacity(0.45))
                        .frame(width: 1)
                        .offset(x: width(Double(tick), in: geo.size.width))
                }
            }
            .clipShape(Capsule())
        }
        .frame(height: height)
        .accessibilityHidden(true)
    }

    private func width(_ percent: Double, in total: CGFloat) -> CGFloat {
        total * CGFloat(min(max(percent, 0), 100)) / 100
    }

    /// A contract tone (`neutral`, `track`, or one of the avatar hues) as a
    /// colour. The hues are the SAME palette the avatars and board icons use,
    /// so a layer's colour means nothing else on screen; `neutral` is the
    /// unopinionated white the agent's own base prompt gets, and `track` is
    /// the rail itself (the `free` legend row's swatch).
    public static func tone(_ tone: String) -> Color {
        switch tone {
        case "neutral": .white.opacity(0.3)
        case "track": GlassTokens.strokeStrong
        case "red": DesignTokens.Avatar.red
        case "orange": DesignTokens.Avatar.orange
        case "yellow": DesignTokens.Avatar.yellow
        case "green": DesignTokens.Avatar.green
        case "teal": DesignTokens.Avatar.teal
        case "blue": DesignTokens.Avatar.blue
        case "violet": DesignTokens.Avatar.violet
        case "pink": DesignTokens.Avatar.pink
        // A tone a newer contract added: the neutral fill, never nothing.
        default: .white.opacity(0.3)
        }
    }
}
